d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -207,7 +259,16 @@ public void onChanged(XdsClient.D2NodeUpdate update)
           {
             ClusterProperties clusterProperties = toClusterProperties(update.getNodeData().getData(),
                 update.getNodeData().getStat().getMzxid());
-            _clusterEventBus.publishInitialize(clusterName, clusterProperties);
+            // For symlink clusters, ClusterLoadBalancerSubscriber subscribed to the symlinks, instead of the actual node, in event bus,
+            // so we need to publish under the symlink names.
+            // For other clusters, publish under its original name. Note that these clusters could be either:
+            // 1) regular clusters requested normally.
+            // 2) clusters that were pointed by a symlink previously, but no longer the case after the symlink points to other clusters.
+            // For case #2: the symlinkAndActualNode map will no longer has an entry for this cluster (removed in
+            // D2SymlinkNodeResourceWatcher::onChanged), thus the updates will be published under the original cluster name
+            // (like "FooCluster-prod-ltx1"), which has no subscribers anyway, so no harm to publish.
+            String clusterNameToPublish = StringUtils.defaultString(getSymlink(clusterName), clusterName);
```

Comment:
For symlink enabled clusters, can the clients never request a specific fabric (eg. `FooCluster-prod-ltx1`)? Looked at one of the symlink-enabled clusters (`$JobServicesMaster`), and the service `/d2/services/jobIngestionBlacklistRules` ZK data instead has `clusterName` set to `JobsServices-prod-ltx1` directly. From the code change, it appears that if `$JobServicesMaster` points to `JobsServices-prod-ltx1` and there's a subscriber for the former, then the latter will never be published. Maybe both should be published instead?
```
{
  "path": "/jobIngestionBlacklistRules",
  "serviceName": "jobIngestionBlacklistRules",
  "transportClientProperties": {
    "http.idleTimeout": "10500000",
    "allowedClientOverrideKeys": "http.protocolVersion,http.requestTimeout",
    "http.poolSize": "200",
    "http.maxHeaderSize": "16834",
    "http.maxResponseSize": "10000000",
    "http.sslIdleTimeout": "10500000",
    "http.requestTimeout": "10000"
  },
  "backupRequests": [],
  "serviceMetadataProperties": {
    "isDefaultService": "true",
    "restli.protocol": "2.0.0"
  },
  "clusterName": "JobsServices-prod-ltx1",
```

Reply:
clients make calls to d2 service jobIngestionBlacklistRulesMaster instead for symlink-enabled service.
That said, users indeed could send calls directly to non "Master" d2 services too. We should publish both. Thanks for the deep thinking, investigation, and nice catch!

Reply:
Thanks, would be good to add a test-case specifically for this too.

Reply:
updated